compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -2319,6 +2319,10 @@ impl<'tcx> Ty<'tcx> {
 
     /// Returns the type of the async destructor of this type.
     pub fn async_destructor_ty(self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>) -> Ty<'tcx> {
+        if self.is_async_destructor_noop(tcx, param_env) || matches!(self.kind(), ty::Error(_)) {
```

Comment:
prefer `self.references_error()` over matching for `ty::Error`

Reply:
Neither drop glue nor discriminant kind do this. I will have to switch every match on `ty::Error` otherwise compiler may emit type errors.

Reply:
I think I cannot add this case to be a part of `is_async_destructor_noop` because of this line:

https://github.com/rust-lang/rust/blob/e239e73a77cda38b53e9a7716e01f0c2af995ff5/compiler/rustc_trait_selection/src/traits/project.rs#L1111

But I'm not sure, maybe I should include it?

Reply:
Hmm... I guess we can look if there's a general thing to improve here everywhere